### ISSUE TITLE
On-Premise HA Support

### ISSUE THREAD

**Issue by moshloop** [kubernetes-sigs/cluster-api-provider-vsphere#461]
With most of the development and testing of CAPV happening on VMC, I think it is important to outline the plan for supporting traditional on-premise workloads that CAPV is intended for.  I would guess that most people who are deploying in AWS will use CAPA and not CAPV on VMC.

**Comment by akutz** [comment]
VMC on AWS does not mean AWS. VMC on AWS is bare metal vSphere on AWS.

**Comment by moshloop** [comment]
VMC means not on-premise - CAPV needs to cater for on-premise environments as that is where 99.9% of vSphere is deployed.

**Comment by akutz** [comment]
Fair. I suggest rewording the problem description to focus on 'on-prem' as the distinction of AWS vs VMC on AWS. FWIW, VMC on AWS is actually marketed as "on-prem in the cloud," so the distinction may not be as clear as you might think it should be.

**Comment by akutz** [comment]
FWIW, I've had conversations with colleagues at VMware germane to this issue. I'm going to copy some of those details below:

> **Overview**
> CAPV does not currently support deploying HA Kubernetes clusters, and VMware's ability to provide a solution for this is limited.
>
> **In scope**
> In the context of this issue, load balancer refers to providing HA access to multiple members of a Kubernetes control plane.
>
> **Out-of-scope**
> This issue is not concerned with providing load balancers to Kubernetes services.
>
> **Details**
> vSphere on-prem lacks functionality often considered standard amongst cloud platforms:
>
> * metadata service
> * load balancer service
>
> Thus CAP* implementations for cloud platforms may not encounter the same challenges, such as lack of a load balancing solution, as CAPV.
>
> Here is what VMware can offer today:
>
> | SKU | IPv4 | IPv6 |
> |-----|------|------|
> | vSphere | | |
> | vSphere+NSX | ✓ | Unclear |
>
> The plan for moving forward is to simply punt automated HA for now, and we will document the following in order to provision Kubernetes clusters with CAPV that support HA:
>
> 1. Assign static IP addresses to machines that are members of the control plane
> 2. One of the following:
>     * Using an external load balancer, ex. Netscaler, set up a round-robin/sticky IP that points to the control plane member IP addresses, or
>     * Create a Round Robin DNS record that points to the control plane member IP addresses
> 3. Create the Kubernetes cluster with the control plane endpoint set to the IP address/FQDN provided by the external load balancer or RR DNS record
>
> Regardless, NSX is not built into all vSphere SKUs, and the question remains – what is the answer/solution/story when asked whether or not CAPV supports deploying HA Kubernetes clusters. Is it predicated on NSX? If so, are the steps to automate standing up an NSX load balancer built into CAPV?

**Comment by akutz** [comment]
Hi @timothysc,

I suggest using this issue to discuss HA support for CAPV moving forward. Please also read the above comments to make sure the scope of this issue is understood. This issue refers to HA access:

* to the control plane endpoints
* from clients external to the cluster

Solutions that provide HA access for the API server for clients inside the cluster or are focused on HA for K8s services are out-of-scope for this issue.
